Use GroupVersion.WithResource

skitt opened the pull request:
This simplifies GroupVersionResource construction code.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

skitt committed:
Use GroupVersion.WithResource

This simplifies GroupVersionResource construction code.

To avoid introducing dependencies on deprecated SchemeGroupVersion
declarations (e.g. in mcsv1a1), add a helper to do a similar job using
GroupVersion.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

tpantelis commented on internal/gather/servicediscovery.go:
```diff
@@ -49,19 +48,11 @@ func gatherCoreDNSPodLogs(info *Info) {
-		Version:  mcsv1a1.GroupVersion.Version,
-		Resource: "serviceexports",
-	}, namespace, metav1.ListOptions{})
+	ResourcesToYAMLFile(info, mcsv1a1.SchemeGroupVersion.WithResource("serviceexports"), namespace, metav1.ListOptions{})
```
`mcsv1a1.SchemeGroupVersion` is actually deprecated - I'm guessing newer versions of the code generator don't generate it but the authors kept it for backwards compatibility. I think we should leave this as is - also in other places.

skitt replied:
They still do, the comment was added following https://github.com/kubernetes/kubernetes/pull/66251#discussion_r203715354 but I don’t know why `SchemeGroupVersion` is supposed to be deprecated (which puts a damper on this whole PR really).

tpantelis replied:
yeah and someone asked that over a year ago but wasn't answered. We still need to use `SchemeGroupVersion` for the ones where the code has not been re-generated using a newer generator that deprecates it. So I think this PR is still valid, ie we might as well use `WithResource` where we can w/o being warned about deprecation.

skitt committed:
Use the OpenShift security GroupVersion

This makes the dependency explicit and removes a manual GVR
declaration (we still need the resource name however).

Signed-off-by: Stephen Kitt <skitt@redhat.com>

tpantelis commented on internal/gvr/resource.go:
```diff
@@ -0,0 +1,32 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+func ResourceFromMetaGroupVersion(gv metav1.GroupVersion, resource string) schema.GroupVersionResource {
```
I think we can shorten this.

That said, I'm planning to change `gather` to use a controller client instead of a dynamic client in which case we would use GVK instead of GVR. Perhaps we can hold off on these changes for now.

```suggestion
func FromGroupVersion(gv metav1.GroupVersion, resource string) schema.GroupVersionResource {
```

tpantelis replied:
I've decided to keep the dynamic client for ease of use so this change is relevant.

skitt replied:
I’ve switched to `FromMetaGroupVersion` instead of `ResourceFrom…` because we’re specifically handling `meta` `GroupVersion` as opposed to `schema`; I’d rather keep the `Meta` reference in case we ever need to do something with one of the other versions.

tpantelis commented on internal/gvr/resource.go:
```diff
@@ -0,0 +1,32 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+func FromMetaGroupVersion(gv metav1.GroupVersion, resource string) schema.GroupVersionResource {
```
It would probably be useful to put this in **admiral** but we can do that later if needed.

dfarrell07 committed:
Merge branch 'devel' into withresource
